---
title: 'SciX: Genies'
author: "Lachlan Gray"
date: "19/06/2022"
output: html_document
runtime: shiny
---
### Important Note 

If you have already successfully installed all packages in a previous script such as SciX.Genies.HD.metadata.1.Rmd you do not need to do so again and can start this script from the "Load the packages" section. 

### Introduction

This file is an Rmarkdown file. Upon successfully installed R and RStudio you should be able to follow the instructions below. If you run into errors please flag on the GitHub page or search the error in Google.

To run each chunk of code press the green arrow next to the command or select the code and run with 'command + enter'.

Opening this Rmarkdown file will place us into the required working directory. A directory is another name for folders on the computer. All of the plots and files we generate will be saved to the working directory.
We can find which directory we are in with the getwd() command:

```{r echo=FALSE}
getwd()
```

For MacOS and Linux getwd() should return the following path:
"/Users/USER/Desktop/SciX-main/HD"

We can then see which files are available in this directory with the dir() command:
```{r setup, include=FALSE}
dir()
```


### Setting up R
Only run this once
```{r setup, include=FALSE}
install.packages('knitr', verbose=F)
knitr::opts_chunk$set(echo = TRUE)
r <- getOption("repos")
r["CRAN"] <- "https://cran.csiro.au/"
options(repos = r)
```

### Install packages 
Only run this once
If asked to update all/some/none just enter 'a' in the console below.
```{r, echo=FALSE}
install.packages('ggplot2', verbose = F)
install.packages('reshape2', verbose = F)
install.packages('ggrepel', verbose = F)
install.packages('ggupset', verbose = F)
install.packages('BiocManager', verbose = F)
install.packages('statmod', verbose = F)
install.packages('pheatmap', verbose = F)
BiocManager::install("qvalue")
BiocManager::install("edgeR")
BiocManager::install("clusterProfiler")
```

### Load the packages
Start from here every time you open RStudio to run this script
```{r echo=FALSE}
library(ggplot2, verbose = F)
library(qvalue, verbose = F)
library(reshape2, verbose = F)
library(ggrepel, verbose = F)
library(ggupset, verbose = F)
library(enrichplot, verbose = F)
library(edgeR, verbose = F)
library(DOSE, verbose = F)
library(clusterProfiler, verbose = F)
library(statmod, verbose=F)
library(pheatmap, verbose=F)
```


### Read in RNA sequencing count matrix. This is the data that we will be using for this experiment
The read.csv command allows us to load a comma separated file into R. This file contains data in the form of a matrix (a grid of numbers). The "header" option is set to "T" which means that the first row of the file contains the names of the columns. The "row.names" option is set to "1" which means that the first column of the file containing the gene names is used to name each row.
We then print the first five rows of the matrix (which includes all of the columns) to the screen. It also prints the dimensions of the matrix, which tells us the number of rows (genes) and columns (individuals) in the matrix.
```{r echo=FALSE}
# Read in comma separated count matrix
count.matrix <- read.csv('HD.data.csv', 
                         header=T, row.names = 1)
# print the first five rows and all of the columns
count.matrix[1:5,]
# Print the dimensions of this matrix. This tells us we have 58,721 genes and 13 individuals
dim(count.matrix)
```


### Read in sample metadata. This contains basic information about each sample.
The read.delim() command is similar to the read.csv() command but is more flexible because we can read in files which use tabs '\t' which separate the columns. Here we replace the sample accession number with a more informative sample ID.
```{r echo=FALSE}
metadata <- read.delim('HD.metadata.txt', sep='\t')
print(metadata)

colnames(count.matrix) <- metadata$condition
# Print first 5 genes
print(count.matrix[1:5,])
```

### Adding biological sex to the metadata file
You may have noticed that information about the individuals age and sex is missing from the metadata. By looking at expression of genes on the X and Y chromosomes we can determine the biological sex of these samples. The method to infer sex from gene expression is a little complicated but we can return to this later if you'd like.
```{r echo=FALSE}
# Load in chrY data from the directory behind the working directory
load('../chrY.Rdata')

# First we infer sex based on expression of female specific XIST gene
XIST.expression <- unlist(count.matrix[grep('XIST', rownames(count.matrix)),])

# Filter matrix for XIST and Y chromosome genes then filter and transpose data to have genes on columns and samples on rows
count.matrix.Y <- count.matrix %>%
  mutate(gene = gsub('.+\\|', '', rownames(count.matrix))) %>%
  filter(gene %in% c('XIST', rownames(chrY))) %>% 
  select(-gene) %>%
  filter(rowSums(.) > nrow(.)) %>% 
  t() %>%
  data.frame()

# Perform hierarchical clustering to identify groups
dissimilarity <- dist(count.matrix.Y[,-1], method='euclidean')
cluster <- hclust(dissimilarity, method = 'centroid')
plot(cluster)
cluster.result <- cutree(cluster, k=5)

# Assign sex based on dendrogram
metadata$sex <- ifelse(cluster.result == 1, 'F', 'M')
print(metadata)
# sanity check of XIST expression
print(XIST.expression)

```

# Differential expression analysis with edgeR likelihood ratio test
We will perform a statistical test to determine which genes are different between our conditions. For this, we will use the likelihood ratio test which takes models from each condition and compares them.
We then make our disease samples the reference group. This tells us the difference in gene expression in relation to our disease group. For example, a gene with a positive (+) logFC is upregulated in disease and a negative (-) logFC is downregulated in disease.
We then filter out lowly expressed genes, normalise the expression values and perform the test.
To visualise our results, we create plots to show differentially expressed genes.

```{r echo=FALSE}
# Create a dataframe containing all of the required information
targets = data.frame(group = gsub('[0-9]+', '', metadata$condition),
                     sex = metadata$sex)
# Create the model which considers the sex and condition in the analysis
design <- model.matrix(~0+sex+group, data=targets)
y = DGEList(counts = count.matrix, group = targets$group)
# Change the reference group to the disease group
y$samples$group <- factor(y$samples$group, levels=c('H', 'C'))
# Filter lowly expressed genes
keep <- filterByExpr(y)
y <- y[keep, ,keep.lib.sizes=FALSE]
# Normalise the data to remove between sample differences
y <- calcNormFactors(y, method = "TMM")
# Transform data into a counts per million value
tmm <- cpm(y)
# Estimate the dispersion in the data
y <- estimateDisp(y, design, robust=TRUE)
# Fit data to a model
fit <- glmQLFit(y, design)
# Likelihood ratio test to compare models
lrt <- glmLRT(fit)
# Return results
tt <- topTags(lrt, n=nrow(y), p.value=0.05)

# Smearplot
plotSmear(lrt, de.tags=rownames(tt$table))
pdf('smearplot.pdf')
plotSmear(lrt, de.tags=rownames(tt$table))
dev.off()

# Print the results
print(summary(decideTests(lrt)))

# Convert pvalues into a False Discovery Rate
lrt <- as.data.frame(lrt)
lrt$FDR <- qvalue(lrt$PValue)$qvalues
# Replace gene names
gene <- gsub('ENSG[0-9]+.[0-9]+\\|', '', rownames(lrt))
lrt <- cbind(gene,lrt)

# Print the top 10 differentially expressed genes
lrt[order(lrt$FDR),][1:10,]

# MDS plot
mds <- plotMDS(y, top=1000, plot=F)
toplot <- data.frame(Dim1 = mds$x, Dim2=mds$y, group=factor(y$samples$group), sex=factor(metadata$sex))
mdsplot <- ggplot(toplot, aes(Dim1, Dim2, colour=group)) + 
  geom_point(aes(shape=sex), size=5) +
  scale_shape_manual(values=c(17, 15)) +
  geom_text_repel(data=toplot, label=rownames(toplot)) +
  ggtitle('Multidimensional scaling (MDS) plot')
mdsplot

pdf('MDS.plot.pdf')
mdsplot
dev.off()
```

### Save result file to working directory
```{r echo=TRUE}
write.table(lrt, row.names = F, sep = "\t", 'edgeR-LRT.HD.txt')
```

## Displaying results in volcano plot
This plot displays the log fold-change and false discovery rate for each gene. You can select the number of genes to label with the **n.genes** variable below.
```{r echo=FALSE}
n.genes = 20

threshold <- lrt$FDR < 0.05 & abs(lrt$logFC) > 0.5
lrt$threshold <- threshold

# Order on FDR
ordered <- lrt[order(lrt$FDR),]
volcano.plot <- ggplot(lrt) +
  geom_point(aes(x=logFC, y=-log10(FDR), colour=threshold)) +
  geom_text_repel(data=head(ordered, n.genes), aes(x=logFC, y=-log10(FDR), label=gene)) +
  ggtitle("Volcano Plot: Huntington's Disease") +
  xlab("log2 fold change") +
  ylab("-log10 FDR") +
  theme(plot.title = element_text(size = rel(1.5), hjust = 0.5),
        axis.title = element_text(size = rel(1.25))) +
  scale_color_discrete(name = "DEG")
volcano.plot

pdf(paste0('Volcano.Plot.top', n.genes, '.pdf'))
volcano.plot
dev.off()
```

# Over Representation Analysis (ORA)
Over-representation analysis is used to determine which defined gene sets are more present (over-represented) in a subset of genes than what would be expected by chance. We can use this to associate lists of DE genes to a possible function. 
Here we are comparing them to the Reactome database, which is a collection of information on biological processes and pathways that also contains information about which genes are involved. This information is collated from functional studies on these processes and pathways. 
In the dotplot generated below, the y axis is the top 20 enriched gene sets, and the x axis of the plot (GeneRatio) refers to the ratio of significant genes to the total number of genes within the set.
Note: if dotplot reactome names are overlapping and unreadable try changing "font.size=12" to a lower number.
In the upset plot the number of genes which overlap multiple gene sets are shown. 
```{r echo=FALSE}
gene.set <- read.gmt('../reactome.gmt')
sigDE <- subset(lrt, abs(logFC) > 0.5 & FDR < 0.05)

ego <- enricher(gene = sigDE$gene, 
         universe = lrt$gene,
         pAdjustMethod = "fdr", 
         qvalueCutoff = 0.01,
         TERM2GENE = gene.set)

dotplot(ego, showCategory=20, orderBy="GeneRatio",font.size=12)
pdf('reactome.ORA.pdf', width = 14)
dotplot(ego, showCategory=20, orderBy="GeneRatio",font.size=12)
dev.off()

upsetplot(ego) + theme(axis.text.y = element_text(size=6))
pdf('reactome.upsetplot.pdf', width = 14)
upsetplot(ego)
dev.off()
```

# Save over-respresented reactome terms to working directory
This saves a file called 'reactome.Terms.list.txt' that can then be opened in excel or another text editor. The line or row number of a pathway is the number to be entered in the "Extract genes from interesting pathway" section lower down to see boxplots of genes in that pathway (only if run directly after this code chunk). A pathway of interest can also be searched at https://reactome.org/PathwayBrowser/ to see what broader biological process it relates to. 
```{r echo = TRUE}
#create a. list of over-represented reactome terms 
ego.df <- data.frame(ego)
#output to a text file we can look at in excel or any text editor 
write.table(ego.df, row.names = F,col.names = F, sep = "\t", 'reactome.Terms.list.txt')
#show first 5 rows
ego.df[1:5,]

```

### Match genes to DisGeNET and perform chi-squared test
DisGeNET is a large database containing associations between genes and diseases from thousands of studies. 
We have downloaded the list of genes associated with Huntington's Disease and we can see if our list of DE genes contains a significant overlap with these genes. 
We will also do another ORA this time with just the genes that we know are significantly differentially expressed AND associated with HD (from DisGeNET).
```{r echo=FALSE}
disgene <- read.delim('disgenenet.HD.tsv', header=T, sep="\t")
disgene <- unique(disgene$Gene)

# calculate numbers for different conditions
# disease-associated and DE 
a <- lrt[lrt$FDR < 0.05 & abs(lrt$logFC) > 0.5 & lrt$gene %in% disgene,]$gene
# not disease-associated and DE 
b <- lrt[lrt$FDR < 0.05 & abs(lrt$logFC) > 0.5 & !(lrt$gene %in% disgene),]$gene
# disease-associated and not DE
c <- lrt[lrt$FDR > 0.05 & lrt$gene %in% disgene,]$gene
# not disease-associated and not DE
d <- lrt[lrt$FDR > 0.05 & !(lrt$gene %in% disgene),]$gene
ctable <- matrix(c(length(a),length(b),length(c),length(d)),nrow=2, ncol=2)
chisq <- chisq.test(ctable)

print('Expected values')
chisq$expected

print('Observed values')
chisq$observed

print('Pearson residuals') #Positive values indicate a positive association
chisq$residuals

print('chi.squared p.value')
chisq$p.value

# ORA of significant genes known to cause disease
ego <- enricher(gene = a, 
                universe = lrt$gene,
                pAdjustMethod = "fdr", 
                qvalueCutoff = 0.01,
                TERM2GENE = gene.set)
dotplot(ego, orderBy="GeneRatio", font.size=10, label_format=20)
pdf('disgenenet.ORA.pdf', width=14)
dotplot(ego, orderBy="GeneRatio", font.size=12, label_format=20)
dev.off()
```

# Save reactome terms identified in DisGeNET 
This saves a file called 'disgenenet.Terms.list.txt' that can then be opened in excel or another text editor, but this file only includes terms with genes identified in DisGeNET and Differentially expressed. The line or row number of a pathway is the number to be entered in the "Extract genes from interesting pathway" section below to see boxplots of genes in that pathway (only if run directly after this code chunk). A pathway of interest can also be searched at https://reactome.org/PathwayBrowser/ to see what broader biological process it relates to.
```{r echo=TRUE}
#create a. list of over-represented reactome terms associated with DisGeNET genes 
ego.df <- data.frame(ego)
#output to a text file we can look at in excel or any text editor 
write.table(ego.df, row.names = F,col.names = F, sep = "\t", 'disgenenet.Terms.list.txt')
#show first 5 rows
ego.df[1:5,]
```

### Extract genes from interesting pathway. Select pathway with **pathway** variable
The number to enter as the **pathway** variable can be found as the line or row number of that pathway/term in the 'reactome.Terms.list.txt' or 'disgenenet.Terms.list.txt' files. This chunk will then plot expression of all genes in that pathway as boxplots.
Note: if boxplot scales are being thrown off by dots(genes) with very high expression try setting "outliers = F" on line 337
```{r echo=FALSE}
pathway <- 1
plot.genes <- unlist(strsplit(data.frame(ego)[pathway,11], '/'))
rownames(tmm) <- gsub('ENSG[0-9]+.[0-9]+\\|', '', rownames(tmm))
colnames(tmm) <- gsub('[0-9]+', '', colnames(tmm))
plot.data <- subset(tmm, rownames(tmm) %in% plot.genes)
plot.data <- plot.data[order(rowSums(plot.data), decreasing = T),]
plot.data <- melt(plot.data)
colnames(plot.data) <- c('gene', 'condition', 'cpm')
boxplot.plot <- ggplot(plot.data, aes(x=gene, y=cpm, fill=condition)) + 
  geom_boxplot(outliers = T) +
  theme(axis.text.x = element_text(angle=45, hjust=1)) +
  ggtitle(paste('', ego[pathway,1])) +
  ggtitle(paste('Expression of', ego[pathway,1], 'genes'))

boxplot.plot

pdf('boxplot.pdf', width=14)
boxplot.plot
dev.off()

# Show logFC of genes in pathway
subset(lrt, gene %in% plot.data$gene)
```

